local-volume/provisioner/pkg/deleter/deleter.go

```diff
@@ -46,60 +52,178 @@ func (d *Deleter) DeletePVs() {
-				deletingLocalPVErr := fmt.Errorf("Error deleting PV %q: %v", name, err.Error())
-				d.RuntimeConfig.Recorder.Eventf(pv, v1.EventTypeWarning, common.EventVolumeFailedDelete, deletingLocalPVErr.Error())
-				continue
-			}
-			glog.Infof("Deleted PV %q", name)
 		}
 	}
 }
 
-func (d *Deleter) cleanupPV(pv *v1.PersistentVolume) error {
+func (d *Deleter) deletePV(pv *v1.PersistentVolume) error {
 	if pv.Spec.Local == nil {
 		return fmt.Errorf("Unsupported volume type")
 	}
 
 	config, ok := d.DiscoveryMap[pv.Spec.StorageClassName]
 	if !ok {
-		return fmt.Errorf("Unknown storage class name %v", pv.Spec.StorageClassName)
+		return fmt.Errorf("Unknown storage class name %s", pv.Spec.StorageClassName)
+	}
+
+	mountPath, err := common.GetContainerPath(pv, config)
+	if err != nil {
+		return err
 	}
 
 	// TODO: Get volType from PV.
 	volType := common.VolumeTypeFile
+	isblk, _ := d.VolUtil.IsBlock(mountPath)
+	if isblk {
+		volType = common.VolumeTypeBlock
+	}
+
+	if d.ProcTable.IsRunning(pv.Name) {
+		// Run in progress, nothing to do,
+		return nil
+	}
+
+	err = d.ProcTable.MarkRunning(pv.Name)
+	if err != nil {
+		return err
+	}
+
+	go d.asyncDeletePV(pv, volType, mountPath, config)
+
+	return nil
+}
+
+func (d *Deleter) asyncDeletePV(pv *v1.PersistentVolume, volType string, mountPath string, config common.MountConfig) {
+	defer d.ProcTable.MarkDone(pv.Name)
+
+	var err error
 	switch volType {
 	case common.VolumeTypeFile:
-		return d.cleanupFileVolume(pv, config)
+		err = d.deleteFilePV(pv, mountPath, config)
 	case common.VolumeTypeBlock:
-		return fmt.Errorf("Not yet implemented")
+		err = d.cleanupBlockPV(pv, mountPath, config)
 	default:
-		return fmt.Errorf("Unexpected volume type %q for deleting path %q", volType, pv.Spec.Local.Path)
+		err = fmt.Errorf("Unexpected volume type %q for deleting path %q", volType, pv.Spec.Local.Path)
 	}
-}
 
-func (d *Deleter) cleanupFileVolume(pv *v1.PersistentVolume, config common.MountConfig) error {
-	specPath := pv.Spec.Local.Path
-	relativePath, err := filepath.Rel(config.HostDir, specPath)
 	if err != nil {
-		return fmt.Errorf("Could not get relative path: %v", err)
+		glog.Error(err)
+		return
+	}
+
+	// Remove API object
+	if err := d.APIUtil.DeletePV(pv.Name); err != nil {
+		// TODO: Does delete return an error if object has already been deleted?
+		deletingLocalPVErr := fmt.Errorf("Error deleting PV %q: %v", pv.Name, err.Error())
+		d.RuntimeConfig.Recorder.Eventf(pv, v1.EventTypeWarning, common.EventVolumeFailedDelete,
+			deletingLocalPVErr.Error())
+		glog.Error(deletingLocalPVErr)
+		return
 	}
 
-	mountPath := filepath.Join(config.MountDir, relativePath)
+	glog.Infof("Deleted PV %q", pv.Name)
+
+}
+
+func (d *Deleter) deleteFilePV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error {
+	// Make absolutely sure here that we are not deleting anything outside of mounted dir
+	if !strings.HasPrefix(mountPath, config.MountDir) {
```

Comment:
Can this check be done outside since it's common to both file and block?

Reply:
Fixed.